### ISSUE TITLE
pplx task fails to complete

### ISSUE THREAD

**Issue by shlomiapel** [Azure/azure-storage-cpp#83]
I have code which performs reads and writes from azure storage.
The code is as follows (simplified):

```
performIo(vector<Io> ios, bool isPageBlob)
{
  vector<pplx::task<RcStatus> > tasks;

  for(Io& io : ios)
  {
        if(io.isRead())
        {
			read(io, isPageBlob, tasks);
        }
        else // write
        {
			write(io, isPageBlob, tasks);
        }
  }

  auto whenAllDone = pplx::when_all(tasks.begin(), tasks.end());
  whenAllDone.then([tasks](vector<RcStatus> results) mutable 
  {
    Do Something...
  }
}

read(Io& io, bool isPageBlob,vector<pplx::task<RcStatus> >& tasks)
{
   u32 sizeBytes = io.len();
   u8* arr = new array(u8, sizeBytes);
   auto stream = pplx::streams::rawptr_stream<unsigned char>::open_ostream(arr, sizeBytes);

   auto postReadTask = ([arr, io, this](pplx::task<void> prevTask) mutable
	{
		RcStatus rs = RC_ERROR;

		try
		{
			prevTask.get(); // just checking iff there is any exception...
                        //	use array
			rs = RC_SUCCESS;
		}
		catch (azure::storage::storage_exception& e)
		{
                     / /Do something
		}
		catch (exception &e)
		{
			// Do Something
		}
		delete (arr);
		unregister(io);
		return rs;
	});

	register(io);
	if (isPageBlob)
	{
		azure::storage::cloud_page_blob blob = get_page_blob_reference(io.blobName);
		auto ioTask = blob.download_range_to_stream_async(stream, io.offsetBytes, io.sizeBytes).then(postReadTask);
		tasks.push_back(ioTask);
	}
	else
	{
		azure::storage::cloud_block_blob blob = get_block_blob_reference(io.blobName);
		auto ioTask = blob.download_range_to_stream_async(stream, io.offsetBytes, io.sizeBytes).then(postReadTask);
		tasks.push_back(ioTask);
	}
}
```
write is identical to read, only performs write.

The problem I encountered is that the calling method (performIos) is blocked on whenAllDone.then().
I see that one task has not completed or failed. simply hanging (i register each io before sending to read\write and unregister when done. i poll every minute to check for pending ios and how long they are pending. last time, a read action was pending for over 10 minutes.
any ideas what can cause this?

The problem is not consistent and happened 3 times so far (out of i assume thousands of operations).

**Comment by JasonYang-MSFT** [comment]
@shlomiapel 
Can you share more details as follows?

1. Which platform are you using? Windows?
2. the version of the SDK you are using.
3. More details about the hang case.

        After it hang for over 10 minutes, does it finally success or fail?

        Can you share a brief size of the data? especially for the failed one.

        Does it hang only on read? or both read and write are possible? 

        hang on page_blob or block_blob?

**Comment by shlomiapel** [comment]
1. windows server 2008 R2.
2. Azure-Storage/2.1.0
3.1 Never ends.
3.2 data varies between several Blocks (512B) to 16MB.
3.3 last time it happened on read from a page blob. not sure this was always the case. happened 3 times so far.

**Comment by JasonYang-MSFT** [comment]
@shlomiapel 
We have replayed the this problem. Currently investigating it.

If you already have any investigation results or understandings on this problem, please feel free to share with us.

**Comment by JasonYang-MSFT** [comment]
Fixed in #89

**Comment by JasonYang-MSFT** [comment]
@shlomiapel 
Thanks for your contribution.
We have released 2.4.0, this problem is fixed in this release.  You can try our newest release if you want.
